docs/DMS/PRIMARY-DATA-STORAGE/README.md

```diff
@@ -0,0 +1,214 @@
+The development team has [executed initial
+experiments](../../../POC-Applications/POC.Databases/) to compare some aspects
+of performance between this structure and the ODS database from the ODS/API
+Platform v7. Similar records were inserted into three tables: `Student`,
+`StudentSchoolAssociation`, and `StudentSectionAssociation`. Running a virtual
+machine with SQL Server 2022 and Windows Defender disabled, large numbers of
+inserts were run and timed for direct comparison of the insert performance. The
+following tables summarize the initial findings with 1 million records:
+
+| Database | Insert Time | Storage (KB) |
+| -------- | ----------- | ------------ |
+| DMS      | 02:49:42    | 571,528      |
+| ODS      | 03:22:34    | 1,268,968    |
+
+The techniques used in these experiments do not provide perfect comparisons, but
+we believe these results are "essentially" comparable. And that comparison is
+very favorable, supporting the hypothesis of high performance with this database
+design.
+
+### Read Operations
+
+The Data Management Service will be built to support one of two models for
+database queries ("GET all" and "GET by search string"):
+
+1. (Preferred) Use a search database _instead_ of the relational database
+   (examples: OpenSearch, Elasticsearch).
+2. (Alternate) Add query tables in the relational database, which support
+   indexing on the searchable fields. There will be one query table per Resource
+   and one table for all Descriptors.
+
+```mermaid
+erDiagram
+
+    Documents {
+        Guid document_uuid PK "API resource id, clustered"
+        Guid referential_id "Extracted document identity"
+        string project_name "Example: Ed-Fi (for DS)"
+        string resource_name "Example: Student"
+        string resource_version "Example: 5.0.0"
+        JSON edfi_doc "The document"
+    }
+    QuerySchoolIndex ||--|| Documents : ""
+    QuerySchoolIndex {
+        Guid document_uuid PK,FK "School document being indexed"
+        string school_id "SchoolId of the school"
+        string name_of_institution "Name of the school"
+        string et_cetera
+    }
+    QueryStudentIndex ||--|| Documents : ""
+    QueryStudentIndex {
+        Guid document_uuid PK,FK "Student document being indexed"
+        string first_name "First name of student"
+        string last_surname "Last surname of student"
+        string et_cetera
+    }
+```
+
+See [Design Options for Data Management Service Data
+Storage](./ALTERNATIVES-DESIGNS.md) for more information.
+
+### Implementation Notes
+
+The proof-of-concept process helped uncover some details that need to guide the
+implementation:
+
+* Compared to `varbinary`, using `nvarchar` for the JSON documents results in
+  higher performance and smaller storage requirements.
+* The _partition key_ should be defined in a SQL function that takes the
+  _primary key_ value as input, so that an implementation can modify the
+  partition key without having to recompile the C# application.
```

Comment:
I was thinking the number-of-partitions would be an app settings type thing, so no need to recompile. If the UUID -> partition key is fully encapsulated in a stored proc, wouldn't that mean the number-of-partition settings would have to be in some config table that gets queried a lot by the stored proc? If it's an app setting, either the C# could derive the partition key and pass it directly as a stored proc parameter, or the number-of-partitions config setting could be a parameter.

Reply:
I think we may also want the partition key computation to be in C# for the query table support. We wouldn't be able to do a simple insert into a query table without knowing the partition key of the document. If it were done in a stored proc, we'd have to get the partition key returned from the DB first. Doable but more complex than just having it to begin with.

Reply:
Thinking about this...